src/gui/qgsnewvectortabledialog.cpp

```diff
@@ -0,0 +1,753 @@
+
+  QgsGui::enableAutoGeometryRestore( this );
+  setWindowTitle( tr( "New Table" ) );
+
+  auto updateTableNames = [ = ]( const QString &schema = QString( ) )
+  {
+    mTableNames.clear();
+    try
+    {
+      const auto constTables { conn->tables( schema ) };
+      for ( const auto &tp : constTables )
+      {
+        mTableNames.push_back( tp.tableName() );
+      }
+      validate();
+    }
+    catch ( QgsProviderConnectionException &ex )
+    {
+      // This should never happen but it's not critical, we can safely continue.
+      QgsDebugMsg( QStringLiteral( "Error retrieving tables from connection: %1" ).arg( ex.what() ) );
+    }
+  };
+
+  mTableName->setText( QStringLiteral( "new_table_name" ) );
+  mFieldsTableView->setModel( mFieldModel );
+  QgsNewVectorTableDialogFieldsDelegate *delegate { new QgsNewVectorTableDialogFieldsDelegate( mConnection->nativeTypes(), this )};
+  mFieldsTableView->setItemDelegate( delegate );
+  mFieldsTableView->setSelectionBehavior( QAbstractItemView::SelectionBehavior::SelectRows );
+  mFieldsTableView->setSelectionMode( QAbstractItemView::SelectionMode::SingleSelection );
+  mFieldsTableView->setVerticalHeader( nullptr );
+
+  // Cosmetics
+  mFieldsTableView->horizontalHeader()->setStretchLastSection( true );
+  mFieldsTableView->setColumnWidth( QgsNewVectorTableFieldModel::ColumnHeaders::Type, 300 );
+
+  // Schema is not supported by all providers
+  if ( mConnection->capabilities().testFlag( QgsAbstractDatabaseProviderConnection::Capability::Schemas ) )
+  {
+    mSchemaCbo->addItems( mConnection->schemas() );
+    connect( mSchemaCbo, qgis::overload<const QString &>::of( &QComboBox::currentIndexChanged ), this, [ = ]( const QString & schema )
+    {
+      updateTableNames( schema );
+    } );
+  }
+  else
+  {
+    mSchemaCbo->hide();
+    mSchemaLabel->hide();
+  }
+
+  // Initial load of table names
+  updateTableNames( mSchemaCbo->currentText() );
+
+  // Validators
+  connect( mTableName, &QLineEdit::textChanged, this, [ = ]( const QString & )
+  {
+    validate();
+  } );
+
+  connect( mGeomColumn, &QLineEdit::textChanged, this, [ = ]( const QString & )
+  {
+    validate();
+  } );
+
+  // Enable/disable geometry options and call validate
+  connect( mGeomTypeCbo, qgis::overload<int>::of( &QComboBox::currentIndexChanged ), this, [ = ]( int index )
+  {
+    const bool hasGeom { index != 0 };
+    mGeomColumn->setEnabled( hasGeom );
+    mGeomColumnLabel->setEnabled( hasGeom );
+    mSpatialIndexChk->setEnabled( hasGeom );
+    mSpatialIndexLabel->setEnabled( hasGeom );
+    mCrs->setEnabled( hasGeom );
+    mCrsLabel->setEnabled( hasGeom );
+    mDimensionsLabel->setEnabled( hasGeom );
+    mHasMChk->setEnabled( hasGeom );
+    mHasZChk->setEnabled( hasGeom );
+    validate();
+  } );
+
+  // Hardcode geometry types
+  // TODO: this information should really come from the connection through the provider but there is no API for that yet
```

Comment:
Good idea -- definitely something we should consider adding in future. We should also consider curved types support too (e.g. via a capability flag for "supportscurves"), and z / m support.

Reply:
Also a flag for "provider is not sensitive to multi/single polygon/line distinction"

Reply:
Is there any database provider in the connections API that would have this flag right now?
IIRC we currently support:
- postgis
- gpkg
- spatialite
- mssql

Reply:
SQL server doesn't care, that's one...

Reply:
How about in QgsAbstractDatabaseProviderConnection:

```
    /**
     * The GeometryColumnCapability enum represents the geomery column features supported by the connection
     * \since QGIS 3.16
     */
    enum GeometryColumnCapability
    {
      Z = 1 << 1,                    //! Supports Z dimension
      M = 1 << 2,                    //! Supports M dimension
      SinglePart = 1 << 3            //! Multi and single part types are distinct geometry types
    };
```

Same question as above: can it be static or we'd expect different flags from different backends?

Reply:
It's going to be static -- it depends on the actual backend's capabilities and QGIS' provider functionality. But it won't vary connection-by-connection for one particular provider.

We'll also need a "supports curves" flag